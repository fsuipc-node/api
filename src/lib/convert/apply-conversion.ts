import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
