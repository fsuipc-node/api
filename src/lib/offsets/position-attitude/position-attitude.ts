import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const positionAttitude: OffsetList = {
  groundElevation: new Offset({
    value: 0x20,
    name: 'groundElevation',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'ground elevation - feet',
    convert: 'Math.round({VAL} * 3.28084 / 256)',
    type: Type.Int32,
    permission: 'r',
  }),
  magVar: new Offset({
    value: 0x2A0,
    name: 'magVar',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: '-ve is west, +ve east. mag to true by adding this, true to mag by substracting this',
    type: Type.Int16,
    convert: '+({VAL}*360/65536).toFixed(2)',
    permission: 'r',
  }),
  gs: new Offset({
    value: 0x2B4,
    name: 'gs',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'ground speed in kt',
    type: Type.Int32,
    convert: 'Math.floor({VAL} * 3600 / 65536 / 1852)',
    permission: 'r',
  }),
  tas: new Offset({
    value: 0x2B8,
    name: 'tas',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'true air speed - knots',
    type: Type.Int32,
    convert: 'Math.floor({VAL} / 128)',
    permission: 'r',
  }),
  ias: new Offset({
    value: 0x2BC,
    name: 'ias',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'indicated air speed - knots',
    type: Type.Int32,
    convert: 'Math.floor({VAL} / 128)',
    permission: 'r',
  }),
  bpa: new Offset({
    value: 0x2C4,
    name: 'bpa',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'barper pole airspeed - knots',
    type: Type.UInt32,
    convert: 'Math.floor({VAL} / 128)',
    permission: 'r',
  }),
  vs: new Offset({
    value: 0x2C8,
    name: 'vs',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'vertical speed - ft/min',
    type: Type.Int32,
    convert: 'Math.floor({VAL} * 60 * 3.28084 / 256)',
    permission: 'r',
  }),
  whiskeyCompass: new Offset({
    value: 0x2CC,
    name: 'whiskeyCompass',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'whiskey compass - degrees',
    type: Type.Double,
    permission: 'r',
  }),
  vsAtTouchdown: new Offset({
    value: 0x30C,
    name: 'vsAtTouchdown',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'vertical speed - m/s',
    type: Type.Int32,
    convert: 'Math.floor({VAL}*60*3.28084/256)',
    permission: 'r',
  }),
  planeOnground: new Offset({
    value: 0x366,
    name: 'planeOnground',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'plane on the ground',
    type: Type.UInt16,
    convert: '!!{VAL}',
    permission: 'r',
  }),
  latitude: new Offset({
    value: 0x560,
    name: 'latitude',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'latitude - -ve for south +ve for north',
    type: Type.Int64,
    convert: '{VAL} * 90 / (10001750 * 65536 * 65536)',
    permission: 'rw',
  }),
  longitude: new Offset({
    value: 0x568,
    name: 'longitude',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'longitude - -ve for west +ve for east',
    type: Type.Int64,
    convert: '{VAL} * 360 / (65536 * 65536 * 65536 * 65536)',
    permission: 'rw',
  }),
  altitude: new Offset({
    value: 0x570,
    name: 'altitude',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'altitude - AGL',
    type: Type.Int64,
    convert: '+({VAL} * 3.28084 / (65536 * 65536)).toFixed(2)',
    permission: 'rw',
  }),
  pitch: new Offset({
    value: 0x578,
    name: 'pitch',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'pitch - 0 for level, -ve for pitch up, +ve for pitch down',
    type: Type.Int32,
    convert: '+({VAL} * 360 / (65536 * 65536)).toFixed(2)',
    permission: 'rw',
  }),
  bank: new Offset({
    value: 0x57C,
    name: 'bank',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'bank - -ve right bank, +ve left bank',
    type: Type.Int32,
    convert: '+({VAL} * 360 / (65536 * 65536)).toFixed(2)',
    permission: 'rw',
  }),
  heading: new Offset({
    value: 0x580,
    name: 'heading',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'heading - TRUE',
    type: Type.UInt32,
    convert: '+({VAL} * 360 / (65536 * 65536)).toFixed(2)',
    permission: 'rw',
  }),
  viewpointLatitude: new Offset({
    value: 0x5B0,
    name: 'viewpointLatitude',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'viewpoint latitude - -ve for south +ve for north',
    type: Type.Int64,
    convert: '{VAL} * 90 / (10001750 * 65536 * 65536)',
    permission: 'r',
  }),
  viewpointLongitude: new Offset({
    value: 0x5B8,
    name: 'viewpointLongitude',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'viewpoint longitude - -ve for west +ve for east',
    type: Type.Int64,
    convert: '{VAL} * 360 / (65536 * 65536 * 65536 * 65536)',
    permission: 'r',
  }),
  viewpointAltitude: new Offset({
    value: 0x5C0,
    name: 'viewpointAltitude',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'viewpoint altitude - AGL',
    type: Type.Int64,
    convert: '+({VAL} * 3.28084 / (65536 * 65536)).toFixed(2)',
    permission: 'r',
  }),
  viewpointPitch: new Offset({
    value: 0x5C8,
    name: 'viewpointPitch',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'pitch - 0 for level, - for pitch up,  for pitch down',
    type: Type.Int32,
    convert: '+({VAL} * 360 / (65536 * 65536)).toFixed(2)',
    permission: 'r',
  }),
  viewpointBank: new Offset({
    value: 0x5CC,
    name: 'viewpointBank',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'viewpoint bank - - right bank, + left bank',
    type: Type.Int32,
    convert: '+({VAL} * 360 / (65536 * 65536)).toFixed(2)',
    permission: 'r',
  }),
  viewpointHeading: new Offset({
    value: 0x5D0,
    name: 'viewpointHeading',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'viewpoint heading - TRUE',
    type: Type.UInt32,
    convert: '+({VAL} * 360 / (65536 * 65536)).toFixed(2)',
    permission: 'r',
  }),
  verticalSpeed: new Offset({
    value: 0x842,
    name: 'verticalSpeed',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'vertical speed - more precise - ft/min - +ve = up',
    type: Type.Int16,
    convert: '+({VAL} * -3.28084).toFixed(2)',
    permission: 'r',
  }),
  gForce: new Offset({
    value: 0x1140,
    name: 'gForce',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'g-force',
    type: Type.Double,
    permission: 'r',
  }),
  gForceTouchDown: new Offset({
    value: 0x11B8,
    name: 'gForceTouchDown',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'g-force touchdown - unknown unit',
    convert: '{VAL} / 625',
    type: Type.Int16,
    permission: 'r',
  }),
  gForceAlt: new Offset({
    value: 0x11BA,
    name: 'gForceAlt',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'g-force alt unit',
    convert: '{VAL} / 625',
    type: Type.Int16,
    permission: 'r',
  }),
  angleOfAttack: new Offset({
    value: 0x11BE,
    name: 'angleOfAttack',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'angle of attack - percent of max',
    convert: '+(100 - (100 * {VAL} / 32767)).toFixed(2)',
    type: Type.Int16,
    permission: 'r',
  }),
  machSpeed: new Offset({
    value: 0x11C6,
    name: 'machSpeed',
    category: OffsetCategory.POSITION_ATTITUDE,
    description: 'mach speed',
    convert: '{VAL} / 20480',
    type: Type.UInt16,
    permission: 'r',
  }),
};
