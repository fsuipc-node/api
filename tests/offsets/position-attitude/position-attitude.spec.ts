import { positionAttitude as offsets } from '@offsets/position-attitude/position-attitude';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - position-attitude/position-attitude', () => {
  const offsetsTestCases = [
    { name: 'groundElevation', value: 18726, expectedResult: 240 },
    { name: 'magVar', value: 38229, expectedResult: 210.00 },
    { name: 'gs', value: 674292, expectedResult: 19 },
    { name: 'tas', value: 257, expectedResult: 2 },
    { name: 'ias', value: 520, expectedResult: 4 },
    { name: 'bpa', value: 390, expectedResult: 3 },
    { name: 'vs', value: 14, expectedResult: 10 },
    { name: 'vsAtTouchdown', value: 13, expectedResult: 9 },
    { name: 'planeOnground', value: 0, expectedResult: false },
    { name: 'planeOnground', value: 1, expectedResult: true },
    { name: 'latitude', value: 2.2329449e+16, expectedResult: 46.78263288720104 },
    { name: 'longitude', value: -3.6582626e+18, expectedResult: -71.39333265196446 },
    { name: 'altitude', value: 392731796979, expectedResult: 300 },
    { name: 'pitch', value: -238609294, expectedResult: -20 },
    { name: 'bank', value: -238609294, expectedResult: -20 },
    { name: 'heading', value: -238609294, expectedResult: -20 },
    { name: 'viewpointLatitude', value: 2.2329449e+16, expectedResult: 46.78263288720104 },
    { name: 'viewpointLongitude', value: -3.6582626e+18, expectedResult: -71.39333265196446 },
    { name: 'viewpointAltitude', value: 392731796979, expectedResult: 300 },
    { name: 'viewpointPitch', value: -238609294, expectedResult: -20 },
    { name: 'viewpointBank', value: -238609294, expectedResult: -20 },
    { name: 'viewpointHeading', value: 2147483648, expectedResult: 180 },
    { name: 'verticalSpeed', value: -46, expectedResult: 150.92 },
    { name: 'gForceTouchDown', value: 1250, expectedResult: 2 },
    { name: 'gForceAlt', value: 1250, expectedResult: 2 },
    { name: 'angleOfAttack', value: 25456, expectedResult: 22.31 },
    { name: 'machSpeed', value: 61440, expectedResult: 3 },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
