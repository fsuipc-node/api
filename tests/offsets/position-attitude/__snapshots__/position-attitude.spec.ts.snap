// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - position-attitude/position-attitude offsets list should have required properties 1`] = `
Object {
  "altitude": Offset {
    "category": "position_attitude",
    "convert": "+({VAL} * 3.28084 / (65536 * 65536)).toFixed(2)",
    "description": "altitude - AGL",
    "length": undefined,
    "mapping": undefined,
    "name": "altitude",
    "permission": "rw",
    "type": 4,
    "value": 1392,
  },
  "angleOfAttack": Offset {
    "category": "position_attitude",
    "convert": "+(100 - (100 * {VAL} / 32767)).toFixed(2)",
    "description": "angle of attack - percent of max",
    "length": undefined,
    "mapping": undefined,
    "name": "angleOfAttack",
    "permission": "r",
    "type": 2,
    "value": 4542,
  },
  "bank": Offset {
    "category": "position_attitude",
    "convert": "+({VAL} * 360 / (65536 * 65536)).toFixed(2)",
    "description": "bank - -ve right bank, +ve left bank",
    "length": undefined,
    "mapping": undefined,
    "name": "bank",
    "permission": "rw",
    "type": 3,
    "value": 1404,
  },
  "bpa": Offset {
    "category": "position_attitude",
    "convert": "Math.floor({VAL} / 128)",
    "description": "barper pole airspeed - knots",
    "length": undefined,
    "mapping": undefined,
    "name": "bpa",
    "permission": "r",
    "type": 6,
    "value": 708,
  },
  "gForce": Offset {
    "category": "position_attitude",
    "convert": undefined,
    "description": "g-force",
    "length": undefined,
    "mapping": undefined,
    "name": "gForce",
    "permission": "r",
    "type": 8,
    "value": 4416,
  },
  "gForceAlt": Offset {
    "category": "position_attitude",
    "convert": "{VAL} / 625",
    "description": "g-force alt unit",
    "length": undefined,
    "mapping": undefined,
    "name": "gForceAlt",
    "permission": "r",
    "type": 2,
    "value": 4538,
  },
  "gForceTouchDown": Offset {
    "category": "position_attitude",
    "convert": "{VAL} / 625",
    "description": "g-force touchdown - unknown unit",
    "length": undefined,
    "mapping": undefined,
    "name": "gForceTouchDown",
    "permission": "r",
    "type": 2,
    "value": 4536,
  },
  "groundElevation": Offset {
    "category": "position_attitude",
    "convert": "Math.round({VAL} * 3.28084 / 256)",
    "description": "ground elevation - feet",
    "length": undefined,
    "mapping": undefined,
    "name": "groundElevation",
    "permission": "r",
    "type": 3,
    "value": 32,
  },
  "gs": Offset {
    "category": "position_attitude",
    "convert": "Math.floor({VAL} * 3600 / 65536 / 1852)",
    "description": "ground speed in kt",
    "length": undefined,
    "mapping": undefined,
    "name": "gs",
    "permission": "r",
    "type": 3,
    "value": 692,
  },
  "heading": Offset {
    "category": "position_attitude",
    "convert": "+({VAL} * 360 / (65536 * 65536)).toFixed(2)",
    "description": "heading - TRUE",
    "length": undefined,
    "mapping": undefined,
    "name": "heading",
    "permission": "rw",
    "type": 6,
    "value": 1408,
  },
  "ias": Offset {
    "category": "position_attitude",
    "convert": "Math.floor({VAL} / 128)",
    "description": "indicated air speed - knots",
    "length": undefined,
    "mapping": undefined,
    "name": "ias",
    "permission": "r",
    "type": 3,
    "value": 700,
  },
  "latitude": Offset {
    "category": "position_attitude",
    "convert": "{VAL} * 90 / (10001750 * 65536 * 65536)",
    "description": "latitude - -ve for south +ve for north",
    "length": undefined,
    "mapping": undefined,
    "name": "latitude",
    "permission": "rw",
    "type": 4,
    "value": 1376,
  },
  "longitude": Offset {
    "category": "position_attitude",
    "convert": "{VAL} * 360 / (65536 * 65536 * 65536 * 65536)",
    "description": "longitude - -ve for west +ve for east",
    "length": undefined,
    "mapping": undefined,
    "name": "longitude",
    "permission": "rw",
    "type": 4,
    "value": 1384,
  },
  "machSpeed": Offset {
    "category": "position_attitude",
    "convert": "{VAL} / 20480",
    "description": "mach speed",
    "length": undefined,
    "mapping": undefined,
    "name": "machSpeed",
    "permission": "r",
    "type": 5,
    "value": 4550,
  },
  "magVar": Offset {
    "category": "position_attitude",
    "convert": "+({VAL}*360/65536).toFixed(2)",
    "description": "-ve is west, +ve east. mag to true by adding this, true to mag by substracting this",
    "length": undefined,
    "mapping": undefined,
    "name": "magVar",
    "permission": "r",
    "type": 2,
    "value": 672,
  },
  "pitch": Offset {
    "category": "position_attitude",
    "convert": "+({VAL} * 360 / (65536 * 65536)).toFixed(2)",
    "description": "pitch - 0 for level, -ve for pitch up, +ve for pitch down",
    "length": undefined,
    "mapping": undefined,
    "name": "pitch",
    "permission": "rw",
    "type": 3,
    "value": 1400,
  },
  "planeOnground": Offset {
    "category": "position_attitude",
    "convert": "!!{VAL}",
    "description": "plane on the ground",
    "length": undefined,
    "mapping": undefined,
    "name": "planeOnground",
    "permission": "r",
    "type": 5,
    "value": 870,
  },
  "tas": Offset {
    "category": "position_attitude",
    "convert": "Math.floor({VAL} / 128)",
    "description": "true air speed - knots",
    "length": undefined,
    "mapping": undefined,
    "name": "tas",
    "permission": "r",
    "type": 3,
    "value": 696,
  },
  "verticalSpeed": Offset {
    "category": "position_attitude",
    "convert": "+({VAL} * -3.28084).toFixed(2)",
    "description": "vertical speed - more precise - ft/min - +ve = up",
    "length": undefined,
    "mapping": undefined,
    "name": "verticalSpeed",
    "permission": "r",
    "type": 2,
    "value": 2114,
  },
  "viewpointAltitude": Offset {
    "category": "position_attitude",
    "convert": "+({VAL} * 3.28084 / (65536 * 65536)).toFixed(2)",
    "description": "viewpoint altitude - AGL",
    "length": undefined,
    "mapping": undefined,
    "name": "viewpointAltitude",
    "permission": "r",
    "type": 4,
    "value": 1472,
  },
  "viewpointBank": Offset {
    "category": "position_attitude",
    "convert": "+({VAL} * 360 / (65536 * 65536)).toFixed(2)",
    "description": "viewpoint bank - - right bank, + left bank",
    "length": undefined,
    "mapping": undefined,
    "name": "viewpointBank",
    "permission": "r",
    "type": 3,
    "value": 1484,
  },
  "viewpointHeading": Offset {
    "category": "position_attitude",
    "convert": "+({VAL} * 360 / (65536 * 65536)).toFixed(2)",
    "description": "viewpoint heading - TRUE",
    "length": undefined,
    "mapping": undefined,
    "name": "viewpointHeading",
    "permission": "r",
    "type": 6,
    "value": 1488,
  },
  "viewpointLatitude": Offset {
    "category": "position_attitude",
    "convert": "{VAL} * 90 / (10001750 * 65536 * 65536)",
    "description": "viewpoint latitude - -ve for south +ve for north",
    "length": undefined,
    "mapping": undefined,
    "name": "viewpointLatitude",
    "permission": "r",
    "type": 4,
    "value": 1456,
  },
  "viewpointLongitude": Offset {
    "category": "position_attitude",
    "convert": "{VAL} * 360 / (65536 * 65536 * 65536 * 65536)",
    "description": "viewpoint longitude - -ve for west +ve for east",
    "length": undefined,
    "mapping": undefined,
    "name": "viewpointLongitude",
    "permission": "r",
    "type": 4,
    "value": 1464,
  },
  "viewpointPitch": Offset {
    "category": "position_attitude",
    "convert": "+({VAL} * 360 / (65536 * 65536)).toFixed(2)",
    "description": "pitch - 0 for level, - for pitch up,  for pitch down",
    "length": undefined,
    "mapping": undefined,
    "name": "viewpointPitch",
    "permission": "r",
    "type": 3,
    "value": 1480,
  },
  "vs": Offset {
    "category": "position_attitude",
    "convert": "Math.floor({VAL} * 60 * 3.28084 / 256)",
    "description": "vertical speed - ft/min",
    "length": undefined,
    "mapping": undefined,
    "name": "vs",
    "permission": "r",
    "type": 3,
    "value": 712,
  },
  "vsAtTouchdown": Offset {
    "category": "position_attitude",
    "convert": "Math.floor({VAL}*60*3.28084/256)",
    "description": "vertical speed - m/s",
    "length": undefined,
    "mapping": undefined,
    "name": "vsAtTouchdown",
    "permission": "r",
    "type": 3,
    "value": 780,
  },
  "whiskeyCompass": Offset {
    "category": "position_attitude",
    "convert": undefined,
    "description": "whiskey compass - degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "whiskeyCompass",
    "permission": "r",
    "type": 8,
    "value": 716,
  },
}
`;
